File: SmartReceipts/Backup/DataExport.m
//
//  DataExport.m
//  SmartReceipts
//
//  Created by Jaanus Siim on 27/05/15.
//  Copyright (c) 2015 Will Baumann. All rights reserved.
//

#import "DataExport.h"
#import "Objective-Zip.h"
#import "Constants.h"
#import "WBPreferences.h"

@interface DataExport ()

@property (nonatomic, copy) NSString *workDirectory;

@end

@implementation DataExport

- (id)initWithWorkDirectory:(NSString *)pathToDirectory {
    self = [super init];
    if (self) {
        _workDirectory = pathToDirectory;
    }
    return self;
}

- (NSString *)execute {
    NSString *zipPath = [self.workDirectory stringByAppendingPathComponent:SmartReceiptsExportName];

    LOGGER_INFO(@"execute at path: %@", zipPath);
    OZZipFile *zipFile = [[OZZipFile alloc] initWithFileName:zipPath mode:OZZipFileModeCreate];
    LOGGER_INFO(@"zipFile legacy32BitMode: %@", zipFile.legacy32BitMode ? @"YES" : @"NO");
    [self appendFileNamed:SmartReceiptsDatabaseName inDirectory:self.workDirectory archiveName:SmartReceiptsDatabaseExportName toZip:zipFile];
    [self appendAllTripFilesToZip:zipFile];
    NSData *preferences = [[WBPreferences xmlString] dataUsingEncoding:NSUTF8StringEncoding];
    [self appendData:preferences zipName:SmartReceiptsPreferencesExportName toFile:zipFile];
    [zipFile close];

    return zipPath;
}

- (void)appendData:(NSData *)data zipName:(NSString *)zipName toFile:(OZZipFile *)zipFile {
    LOGGER_INFO(@"appendData: size %li, zipName: %@, toFile %@", data.length, zipName, zipFile);
    OZZipWriteStream *stream = [zipFile writeFileInZipWithName:zipName compressionLevel:OZZipCompressionLevelDefault];
    [stream writeData:data];
    [stream finishedWriting];
}

- (void)appendAllTripFilesToZip:(OZZipFile *)file {
    NSString *tripsFolder = [self.workDirectory stringByAppendingPathComponent:SmartReceiptsTripsDirectoryName];
    NSError *error;
    NSArray *trips = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:tripsFolder error:&error];
    if (error) {
        LOGGER_ERROR(@"appendAllTripFilesToZip error: %@", error.description);
    }
    
    for (NSString *trip in trips) {
        [self appFilesForTrip:trip toZip:file];
    }
}

- (void)appFilesForTrip:(NSString *)trip toZip:(OZZipFile *)zip {
    NSString *tripFolder = [[self.workDirectory stringByAppendingPathComponent:SmartReceiptsTripsDirectoryName] stringByAppendingPathComponent:trip];
    NSError *error;
    NSArray *files = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:tripFolder error:&error];
    if (error) {
        LOGGER_ERROR(@"appFilesForTrip:toZip: error: %@", error.description);
    }
    
    for (NSString *file in files) {
        NSString *zipName = [trip stringByAppendingPathComponent:file];
        [self appendFileNamed:file inDirectory:tripFolder archiveName:zipName toZip:zip];
    }
}

- (void)appendFileNamed:(NSString *)fileName inDirectory:(NSString *)containingDirectory archiveName:(NSString *)archiveName toZip:(OZZipFile *)zipFile {
    NSString *filePath = [containingDirectory stringByAppendingPathComponent:fileName];
    NSData *fileData = [NSData dataWithContentsOfFile:filePath];
    [self appendData:fileData zipName:archiveName toFile:zipFile];
}

@end

File: SmartReceipts/Backup/DataExport.swift
//
//  DataExport.swift
//  SmartReceipts
//
//  Created by Bogdan Evsenev on 11/06/2017.
//  Copyright © 2017 Will Baumann. All rights reserved.
//

import Zip

class DataExport: NSObject {
    var workDirectory: String!
    
    init(workDirectory: String) {
        super.init()
        self.workDirectory = workDirectory
    }
    
    func execute() -> String {
        let zipPath = workDirectory.asNSString.appendingPathComponent(backupName)
        
        Logger.debug("Execute at path: \(zipPath)")
        
        // Trips
        var files = tripsFiles()
        
        // DB
        let dbPath = workDirectory.asNSString.appendingPathComponent(SmartReceiptsDatabaseName)
        let dbData = try! Data(contentsOf: dbPath.asFileURL)
        let dbExportPath = workDirectory.asNSString.appendingPathComponent(SmartReceiptsDatabaseExportName)
        _ = FileManager.forceWrite(data: dbData, to: dbExportPath)
        files.append(dbExportPath.asFileURL)
        
        // Preferences
        let preferences = WBPreferences.xmlString().data(using: .utf8)
        let prefPath = workDirectory.asNSString.appendingPathComponent(SmartReceiptsPreferencesExportName)
        let prefExportURL = prefPath.asNSString.deletingLastPathComponent.asFileURL
        _ = FileManager.forceWrite(data: preferences!, to: prefPath)
        files.append(prefExportURL)
        
        do {
            try Zip.zipFiles(paths: files, zipFilePath: zipPath.asFileURL, password: nil, progress: nil)
            try FileManager.default.removeItem(at: dbExportPath.asFileURL)
            try FileManager.default.removeItem(at: prefExportURL)
        } catch {
            Logger.debug("Can't Zip files: \(zipPath)")
        }
        return zipPath
    }
    
    func tripsFiles() -> [URL] {
        let tripsFolder = workDirectory.asNSString.appendingPathComponent(SmartReceiptsTripsDirectoryName)
        var files = [URL]()
        do {
            let trips = try FileManager.default.contentsOfDirectory(atPath: tripsFolder)
            for trip in trips {
                if trip.contains(".SMR") { continue }
                files.append(tripsFolder.asNSString.appendingPathComponent(trip).asFileURL)
            }
        } catch {
            Logger.error("appendAllTripFilesToZip error: \(error.localizedDescription)")
        }
        return files
    }
    
    var backupName:  String {
        let df = DateFormatter()
        df.dateFormat = "YYYY_MM_dd_"
        return df.string(from: Date()) + SmartReceiptsExportName
    }
    
    class func zipFiles(_ urls: [URL], to: String) throws {
        try Zip.zipFiles(paths: urls, zipFilePath: to.asFileURL, password: nil, progress: nil)
    }
}
